internal/spec/spec.go

```diff
@@ -0,0 +1,156 @@
+	spec := &extraSpecs{}
+
+	if len(data.ExtraSpecs) > 0 {
+		if err := json.Unmarshal(data.ExtraSpecs, spec); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal extra specs: %w", err)
+		}
+	}
+
+	return spec, nil
+}
+
+type extraSpecs struct {
+	DiskSize     int64  `json:"disksize,omitempty"`
+	NetworkID    string `json:"network_id,omitempty"`
+	SubnetworkID string `json:"subnetwork_id,omitempty"`
+	NicType      string `json:"nic_type,omitempty"`
+}
+
+func GetRunnerSpecFromBootstrapParams(cfg *config.Config, data params.BootstrapInstance, controllerID string) (*RunnerSpec, error) {
+	tools, err := util.GetTools(data.OSType, data.OSArch, data.Tools)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tools: %s", err)
+	}
+
+	extraSpecs, err := newExtraSpecsFromBootstrapData(data)
+	if err != nil {
+		return nil, fmt.Errorf("error loading extra specs: %w", err)
+	}
+
+	spec := &RunnerSpec{
+		Zone:            cfg.Zone,
+		Tools:           tools,
+		BootstrapParams: data,
+		NetworkID:       cfg.NetworkID,
+		SubnetworkID:    cfg.SubnetworkID,
+		ControllerID:    controllerID,
+		NicType:         defaultNicType,
+		DiskSize:        defaultDiskSizeGB,
+	}
+
+	spec.MergeExtraSpecs(extraSpecs)
+
+	return spec, nil
+}
+
+type RunnerSpec struct {
+	Zone            string
+	Tools           params.RunnerApplicationDownload
+	BootstrapParams params.BootstrapInstance
+	NetworkID       string
+	SubnetworkID    string
+	ControllerID    string
+	NicType         string
+	DiskSize        int64
+}
+
+func (r *RunnerSpec) MergeExtraSpecs(extraSpecs *extraSpecs) {
+	if extraSpecs.NetworkID != "" {
+		r.NetworkID = extraSpecs.NetworkID
+	}
+	if extraSpecs.SubnetworkID != "" {
+		r.SubnetworkID = extraSpecs.SubnetworkID
+	}
+	if extraSpecs.DiskSize > 0 {
+		r.DiskSize = extraSpecs.DiskSize
+	}
+	if extraSpecs.NicType != "" {
+		r.NicType = extraSpecs.NicType
+	}
+}
+
+func (r *RunnerSpec) Validate() error {
+	if r.Zone == "" {
+		return fmt.Errorf("missing zone")
+	}
+	if r.NetworkID == "" {
+		return fmt.Errorf("missing network id")
+	}
+	if r.SubnetworkID == "" {
+		return fmt.Errorf("missing subnetwork id")
+	}
+	if r.ControllerID == "" {
+		return fmt.Errorf("missing controller id")
+	}
+	if r.NicType == "" {
+		return fmt.Errorf("missing nic type")
+	}
+
+	return nil
+}
+
+func (r RunnerSpec) ComposeUserData() (string, error) {
+	switch r.BootstrapParams.OSType {
+	case params.Linux:
+		udata, err := cloudconfig.GetRunnerInstallScript(r.BootstrapParams, r.Tools, r.BootstrapParams.Name)
+		if err != nil {
+			return "", fmt.Errorf("failed to generate userdata: %w", err)
+		}
+
+		lines := strings.Split(string(udata), "\n")
+		if len(lines) > 0 && strings.HasPrefix(lines[0], "#!") {
+			additionalCommands := []string{
+				// Create user 'runner' if it doesn't exist; '|| true' to ignore if user already exists
+				"sudo useradd -m " + defaults.DefaultUser + " || true",
```

Comment:
You also need to make sure the new user is a part of the sudo/admin group.

Reply:
Added the commands to:
- add the `runner` user to the sudo group
- install curl and tar